tools/chpl-language-server/src/chpl-language-server.py

```diff
@@ -404,6 +405,53 @@ def __iter__(self):
         return iter(self.uses)
 
 
+@dataclass
+class EndMarkerPattern:
+    pattern: Union[Type, Set[Type]]
+    header_location: Callable[[chapel.AstNode], Optional[chapel.Location]]
+    goto_location: Callable[[chapel.AstNode], Optional[chapel.Location]]
+
+    @classmethod
+    def all(cls) -> Dict[str, "EndMarkerPattern"]:
+        return {
+            "loop": EndMarkerPattern(
+                chapel.Loop,
+                lambda node: (
+                    node.header_location()
+                    if node.block_style() != "implicit"
+                    else None
+                ),
+                lambda _: None,
```

Comment:
last adjustment, what do you think about writing `pattern = chapel.Loop` etc? I think it might be clearer that way.